### PR TITLE
[wip] Replace `raw.py` with `flyteidl_rust.RawSynchronousFlyteClient`

### PR DESCRIPTION
**Pull request by austin362667**
## Tracking issue

https://github.com/flyteorg/flyte/issues/5344

## Why are the changes needed?

The end game is replacing `raw.py` with fully functioned, auth-included, Pyhton binding exposed from Rust implementation of `RawSynchronousFlyteClient` through PyO3.

With the helps of works in https://github.com/flyteorg/flyte/pull/5525, this PR eliminates the protobuf objects serialization overhead while calling gRPC services. Unlike the original approach in [flyrs](https://github.com/flyteorg/flytekit/tree/flyrs), this method optimizes performance and reduces the maintenance cost.

With [`flyteidl-rust`(test release)](https://test.pypi.org/project/flyteidl-rust/) Python bindings, Flytekit can get rid of not only grpc but also protobuf Python dependencies.

## What changes were proposed in this pull request?

Retain model layer overhead from this PR https://github.com/flyteorg/flytekit/pull/2536, and make current `SynchronousFlyteClient` directly inherit class `RawSynchronousFlyteClient` from binding https://test.pypi.org/project/flyteidl-rust/.

1. Phase 0
- [X] [model] Remove minimal `flyteidl` from model files
- [X] [integration tests] Tests on flyte cluster with public `flytekit` image
- [X] [grpc] Support secured channel in Rust
- [X] [grpc] Catch gRPC error in Python
2. Phase 1
- [X] [model] Remove all `flyteidl` from model files
- [ ] [unit tests] Add Python and Rust tests for `ParseFromString`, `SerializeToString` and `ParseStruct` `ParseValue`…
- [X] [docker] Add flytekit base image with `flyteidl-rust`
- [X] [unit tests] Add `@pytest.mark.remote_task` to skip tests
- [ ] [auth] Oauth in Rust
    - [X] `pkce`
    - [X] `client credentials`
    - [ ] `device flow`
- [X] [integration tests] Tests on flyte cluster with `flyteidl-rust` image
3. Phase 2
- [X] [CI] Publish flyteidl-rust PyPI package
- [ ] [union] Test flyteidl-rust in union sdk
- [X] [project.toml] Remove grpcio and grpcio-status
- [X] [unit tests] Update that use flyteidl
4. Phase 3 (optional)
- [ ] [grpc] Add retry middleware
        Interceptors are a bit more [constrained](https://www.reddit.com/r/rust/comments/11rrwee/tower_middleware_or_interceptor/) in what they can do but are implemented as middleware and requests are nit clonable in the retry loop.
- [ ] [perf] Upgrade [deprecated](https://pyo3.rs/v0.22.1/memory) `gil-refs`

#### Two ways to install flyteidl-rust:
1. Local Generate Protobuf & Build Python Wheel
    1. `git clone -b flyrs https://github.com/austin362667/flyte.git`
    2. `cd flyte/flyteidl/`
    3. `rm -rf gen/pb_rust/* && cargo run --bin gen_flyteidl`
    4. `maturin dev --release -m Cargo.toml`
2. Install flyteidl-rust from pypi test registry
    1. `pip install -i https://test.pypi.org/simple/ flyteidl-rust`

#### Use flytekit with flyteidl-rust:
1. Checkout to dev branch
    1. `git clone -b austin362667/flyrs-v2 https://github.com/austin362667/flytekit.git`
2. Build flytekit image with flyteidl-rust
    1. `docker build -f Dockerfile.rust --build-arg PYTHON_VERSION=3.12 -t localhost:30000/flytekit:rust .`
    2. `docker push localhost:30000/flytekit:rust`
3. Tests
    1. `export FLYTEKIT_IMAGE=localhost:30000/flytekit:rust`
    2. `make integration_test`

### Setup process

0. Checkout to branch `austin362667:austin362667/flyrs-v2`
1. Remove gRPC dependencies
    `pip uninstall grpcio grpcio-status`
6. Install `flyteidl-rust`
    `pip install -i https://test.pypi.org/simple/ flyteidl-rust`
7. Test unauthenticated remote client with gRPC and flyteidl from Rust bindings.
    `pytest ./tests/flytekit/integration/remote/test_remote.py`

### Screenshots

1. PyPi (test): Screenshot 2024-07-24 at 2 01 13 AM

2. Integration Test: Screenshot 2024-07-24 at 2 01 35 AM

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [X] All commits are signed-off.

## Related PRs

https://github.com/flyteorg/flytekit/pull/2536

### CODE REVIEW

File: flytekit/clients/friendly.py [hunk 1]
```diff
@@ -29,7 +28,7 @@
 from flytekit.models.core import identifier as _identifier
 
 
-class SynchronousFlyteClient(_RawSynchronousFlyteClient):
+class SynchronousFlyteClient(flyteidl.RawSynchronousFlyteClient):
```
Comment (austin362667) [inline flytekit/clients/friendly.py:31]: The key modification

File: flytekit/clients/friendly.py [hunk 2]
```diff
@@ -1010,14 +1010,15 @@ def get_upload_signed_url(
...
+                flyteidl.service.CreateUploadLocationRequest(
                     project=project,
                     domain=domain,
                     content_md5=content_md5,
                     filename=filename,
                     expires_in=expires_in_pb,
-                    filename_root=filename_root,
+                    filename_root=filename_root or "",
                     add_content_md5_metadata=add_content_md5_metadata,
+                    org="",
```
Comment (austin362667) [inline flytekit/clients/friendly.py:1021]: initialize `org`  as empty string, can be assigned later by union serverless sdk

File: flytekit/models/execution.py [hunk 3]
```diff
@@ -20,6 +17,13 @@
 from flytekit.models.node_execution import DynamicWorkflowNodeMetadata
 
 
+# A helping function for `from_flyte_idl()`
```
Comment (austin362667) [inline flytekit/models/execution.py:20]: A helping function for `from_flyte_idl()`

File: flytekit/models/execution.py [hunk 4]
```diff
@@ -299,23 +309,31 @@ def to_flyte_idl(self):
...
+            if isinstance(self.notifications, flyteidl.execution_spec.NotificationOverrides.Notifications)
+            else flyteidl.execution_spec.NotificationOverrides.DisableAll(self.disable_all),
             labels=self.labels.to_flyte_idl(),
             annotations=self.annotations.to_flyte_idl(),
             auth_role=self._auth_role.to_flyte_idl() if self.auth_role else None,
             raw_output_data_config=self._raw_output_data_config.to_flyte_idl()
             if self._raw_output_data_config
             else None,
-            max_parallelism=self.max_parallelism,
+            max_parallelism=self.max_parallelism or 0,
```
Comment (austin362667) [inline flytekit/models/execution.py:327]: Is setting to `0` ok?

File: flytekit/models/literals.py [hunk 5]
```diff
@@ -818,16 +829,22 @@ def to_flyte_idl(self):
...
-            error=self.error.to_flyte_idl() if self.error is not None else None,
-            generic=self.generic,
-            structured_dataset=self.structured_dataset.to_flyte_idl() if self.structured_dataset is not None else None,
+        return flyteidl.literal.Value.Scalar(
+            flyteidl.core.Scalar(
+                flyteidl.scalar.Value.Primitive(
+                    self.primitive.to_flyte_idl() if self.primitive is not None else None,
+                    # TODO:
+                    # primitive=self.primitive.to_flyte_idl() if self.primitive is not None else None,
+                    # blob=self.blob.to_flyte_idl() if self.blob is not None else None,
```
Comment (austin362667) [inline flytekit/models/literals.py:838]: TODO

File: flytekit/models/literals.py [hunk 6]
```diff
@@ -929,12 +946,13 @@ def to_flyte_idl(self):
...
-        return _literals_pb2.Literal(
-            scalar=self.scalar.to_flyte_idl() if self.scalar is not None else None,
-            collection=self.collection.to_flyte_idl() if self.collection is not None else None,
-            map=self.map.to_flyte_idl() if self.map is not None else None,
-            hash=self.hash,
-            metadata=self.metadata,
+        return flyteidl.core.Literal(
+            value=self.scalar.to_flyte_idl() if self.scalar is not None else None,
+            # scalar=self.scalar.to_flyte_idl() if self.scalar is not None else None,
+            # collection=self.collection.to_flyte_idl() if self.collection is not None else None,
```
Comment (austin362667) [inline flytekit/models/literals.py:952]: TODO

File: flytekit/models/security.py [hunk 7]
```diff
@@ -162,21 +163,22 @@ def __post_init__(self):
...
-    def from_flyte_idl(cls, pb2_object: _sec.SecurityContext) -> "SecurityContext":
+    def from_flyte_idl(cls, pb2_object: flyteidl.core.SecurityContext) -> "flyteidl.core.SecurityContext":
+        # TODO:
         return cls(
-            run_as=Identity.from_flyte_idl(pb2_object.run_as)
-            if pb2_object.run_as and pb2_object.run_as.ByteSize() > 0
-            else None,
-            secrets=[Secret.from_flyte_idl(s) for s in pb2_object.secrets] if pb2_object.secrets else None,
-            tokens=[OAuth2TokenRequest.from_flyte_idl(t) for t in pb2_object.tokens] if pb2_object.tokens else None,
+            run_as=None,  # Identity.from_flyte_idl(pb2_object.run_as)
```
Comment (austin362667) [inline flytekit/models/security.py:179]: TODO

File: flytekit/models/task.py [hunk 8]
```diff
@@ -479,19 +482,20 @@ def to_flyte_idl(self):
...
+            custom=_json_format.Parse(_json.dumps(self.custom), flyteidl.wkt.Struct()) if self.custom else None,
             task_type_version=self.task_type_version,
             security_context=self.security_context.to_flyte_idl() if self.security_context else None,
             extended_resources=self.extended_resources,
             config={k: v for k, v in self.config.items()} if self.config is not None else None,
-            k8s_pod=self.k8s_pod.to_flyte_idl() if self.k8s_pod else None,
-            sql=self.sql.to_flyte_idl() if self.sql else None,
+            target=flyteidl.task_template.Target.Container(self.container.to_flyte_idl() or None),
+            # container=self.container.to_flyte_idl() if self.container else None,
+            # k8s_pod=self.k8s_pod.to_flyte_idl() if self.k8s_pod else None,
```
Comment (austin362667) [inline flytekit/models/task.py:497]: TODO

File: flytekit/models/types.py [hunk 9]
```diff
@@ -358,23 +356,24 @@ def to_flyte_idl(self):
...
-            metadata=metadata,
+        # if self.metadata is not None:
+        #     metadata = _json_format.Parse(json.dumps(self.metadata), flyteidl.wkt.Struct())
+        # else:
+        #     metadata = None
+
+        t = flyteidl.core.LiteralType(
+            type=flyteidl.literal_type.Type.Simple(int(str(self.simple))) if self.simple else None,
+            # simple=self.simple if self.simple is not None else None,
+            # schema=self.schema.to_flyte_idl() if self.schema is not None else None,
```
Comment (austin362667) [inline flytekit/models/types.py:367]: TODO

File: flytekit/models/types.py [hunk 10]
```diff
@@ -388,24 +387,24 @@ def from_flyte_idl(cls, proto):
...
-            enum_type=_core_types.EnumType.from_flyte_idl(proto.enum_type) if proto.HasField("enum_type") else None,
-            union_type=UnionType.from_flyte_idl(proto.union_type) if proto.HasField("union_type") else None,
-            structured_dataset_type=StructuredDatasetType.from_flyte_idl(proto.structured_dataset_type)
-            if proto.HasField("structured_dataset_type")
-            else None,
-            metadata=_json_format.MessageToDict(proto.metadata) or None,
-            structure=TypeStructure.from_flyte_idl(proto.structure) if proto.HasField("structure") else None,
-            annotation=TypeAnnotationModel.from_flyte_idl(proto.annotation) if proto.HasField("annotation") else None,
+            simple=proto.type if isinstance(proto.type, flyteidl.literal_type.Type.Simple) else None,
+            # schema=SchemaType.from_flyte_idl(proto.schema) if proto.HasField("schema") else None,
```
Comment (austin362667) [inline flytekit/models/types.py:396]: TODO

File: flytekit/clients/friendly.py (hunk 1 above)
Comment (wild-endeavor) [reply flytekit/clients/friendly.py:31]: the other key modification needed is remove these two lines https://github.com/flyteorg/flytekit/blob/a7285067bfd90d53c373163c7db961fe09e3231d/pyproject.toml#L27-L28

Comment (austin362667) [reply flytekit/clients/friendly.py:31]: Yep sure of course, not only "grpcio", "grpcio-status", but also "flyteidl"

File: flytekit/models/core/workflow.py [hunk 11]
```diff
@@ -778,13 +801,19 @@ def on_failure(self):
...
+
     def to_flyte_idl(self):
         """
         :rtype: flyteidl.core.workflow_pb2.WorkflowMetadata
         """
-        workflow_metadata = _core_workflow.WorkflowMetadata()
-        if self.on_failure:
-            workflow_metadata.on_failure = self.on_failure
+
+        workflow_metadata = flyteidl.core.WorkflowMetadata(self.on_failure or 0, self.tags or {})
```
Comment (austin362667) [inline flytekit/models/core/workflow.py:816]: Is setting `on_failure` default value to `flyteidl.workflow_metadata.OnFailurePolicy.FailImmediately` ok?

File: flytekit/models/core/workflow.py [hunk 12]
```diff
@@ -778,13 +801,19 @@ def on_failure(self):
...
+        """
+        :rtype: flytekit.models.core.workflow.WorkflowMetadata.Tags
+        """
+        return self._tags
+
     def to_flyte_idl(self):
         """
         :rtype: flyteidl.core.workflow_pb2.WorkflowMetadata
         """
-        workflow_metadata = _core_workflow.WorkflowMetadata()
```
Comment (austin362667) [inline flytekit/models/core/workflow.py:785]: `WorkflowMetadata` can't accept optional arguments (see Rust flyteidl, there's no Option<T>).
So I use zero and {} as default constructor.

I think it's better to assign default values in Python side for cases like this. That is, we don't try to modify the function signature in Rust.

**Review by pingsutw — APPROVED** [review]

**Pull request merged**